### PR TITLE
[Dev/124] Exam Scoreboard with Barchart

### PR DESCRIPTION
**Pull request by nyngwang**

### CODE REVIEW

File: src/app/exam/questions/ExamQuestions.js [hunk 1]
```diff
@@ -12,95 +12,96 @@ import {useExamContext} from "./ExamContext";
...
+    const {refetchExam, currentExam} = useExamContext()
+    const {examId} = useParams()
+    const NOT_SET = -1
+    const [examQuestions, setExamQuestions] = useState(undefined)
+    const [showAddQuestionModal, setShowAddQuestionModal] = useState(false)
+    const [showRejudgeQuestionModal, setShowRejudgeQuestionModal] = useState(NOT_SET)
+    const [showDeleteQuestionModal, setShowDeleteQuestionModal] = useState(undefined)
+    const [rejudgingProblemId, setRejudgeProblemId] = useState(NOT_SET)
+    const [showEditQuestionModal, setShowEditQuestionModal] = useState(false)
+    const [editingQuestion, setEditingQuestion] = useState(null)
```
Comment (Wally5077) [inline src/app/exam/questions/ExamQuestions.js:33]: 上面是 undefined 下面是 null，這邊是為什麼不同？

File: src/app/exam/questions/ExamQuestions.js [hunk 2]
```diff
@@ -12,95 +12,96 @@ import {useExamContext} from "./ExamContext";
...
-            setQuestions(exam.questions);
-        });
-    }, [examId]);
+    const {url: currentURL} = useRouteMatch()
+    const {refetchExam, currentExam} = useExamContext()
+    const {examId} = useParams()
+    const NOT_SET = -1
+    const [examQuestions, setExamQuestions] = useState(undefined)
+    const [showAddQuestionModal, setShowAddQuestionModal] = useState(false)
+    const [showRejudgeQuestionModal, setShowRejudgeQuestionModal] = useState(NOT_SET)
```
Comment (Wally5077) [inline src/app/exam/questions/ExamQuestions.js:29]: 這邊初始值為什麼是 NOT_SET 不是 false？

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
1. Compose services into a new powerful one that acts as an API Gateway. Don't spread them into view files.
2. Return a powerful model class from the API Gateway to encapsulate domain knowledge (A ScoreBoard class which will compute all statistic functions for you)

File: src/app/exam/score/ExamSummary.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    averageScore,
+    totalExaminees
+ }) => {
+    return <>
+        <div style={{
+            display: "flex",
+            flexDirection: vertical? "column": "row",
+            justifyContent: "flex-start"}}>
+            <Card title="Average Score">
+                <span style={{fontSize: "40px"}}>{averageScore}</span> / 100
```
Comment (Wally5077) [inline src/app/exam/score/ExamSummary.js:15]: NIT: 後面為什麼有一個 / 100？

File: src/app/commons/Card.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import React from "react";
+
+
+const Card = ({
+    title,
+    children,
+    color="#0B5286"
```
[the remark this answers is not in the document]
Comment (Johnny850807) [reply src/app/commons/Card.js:7]: move it into exam's package, as it is uncommon

File: src/app/exam/score/ExamScore.js [hunk 5]
```diff
@@ -1,58 +1,64 @@
...
 import {Spinner} from "../../commons/Spinner";
-import {EmptyCell, TableCell} from "../../../utils/TableCell";
+import {TableCell} from "../../../utils/TableCell";
+import ExamSummary from "./ExamSummary";
+import {recordCollector} from "../../commons/ExamRecordPreprocessing";
+import {examService} from "../../../services/services";
 
 
+// TODO: What if there are more than one problem in this exam?
+//       Current UI design cannot handle this issue.
```
Comment (Chaoyuuu) [inline src/app/exam/score/ExamScore.js:15]: What did you mean?

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

**Review by Chaoyuuu — CHANGES_REQUESTED** [review]
review done

File: src/app/exam/score/ExamSummary.js (hunk 3 above)
Comment (Chaoyuuu) [reply src/app/exam/score/ExamSummary.js:15]: Total score can't be a fixed number (e.g., 100)
It should be the summation of all questions' scores.

File: src/app/exam/live_submissions/ExamSubmissionErrors.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+
+export class UnknownVerdictSummaryError extends Error {
```
[the remark this answers is not in the document]
Comment (nyngwang) [reply src/app/exam/live_submissions/ExamSubmissionErrors.js:2]: Will be removed XD

File: src/app/commons/ExamRecordPreprocessing.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    return {
+        problemId,
+        studentId,
+        judges: judges.map(({testcaseName, grade, maxGrade}) => ({testcaseName, grade, maxGrade})),
+        maxGrade,
+        totalGrade
+    }
+}
+
+export { recordCollector }
```
[the remark this answers is not in the document]
Comment (nyngwang) [reply src/app/commons/ExamRecordPreprocessing.js:13]: Will be moved to `exam/score/`

File: src/app/exam/participants/Examinees.js [hunk 8]
```diff
@@ -10,75 +10,59 @@ import {TextareaModal} from "../../commons/modals/TextareaModal";
...
+                    }
+                },
+            ]}/>
+        </>
     }
 
-    if (!currentExam) {
+    if (!currentExam || !examinees) {
         return <Spinner/>
     }
```
[the remark this answers is not in the document]
Comment (nyngwang) [reply src/app/exam/participants/Examinees.js:67]: This is a bug, thanks

File: src/app/exam/ExamList.js [hunk 9]
```diff
@@ -27,28 +28,61 @@ export const useExamList = () => {
...
 
 const ExamList = () => {
-    const [showCreateExamModal, setShowCreateExamModal] = useState(false);
-    const {exams, setExams} = useExamList();
-    const [currentExam, setCurrentExam] = useState(undefined);
-    const [shouldRedirect, setShouldRedirect] = useState(false);
+    const [showCreateExamModal, setShowCreateExamModal] = useState(false)
+    const {exams, setExams} = useExamList()
+    const [currentExam, setCurrentExam] = useState(undefined)
+    const [currentExaminees, setCurrentExaminees] = useState(undefined)
```
[the remark this answers is not in the document]
Comment (nyngwang) [reply src/app/exam/ExamList.js]: 原本是透過`undefined`來判斷是否初始化, 如果是`[]`則`if (currentExaminees)`會為`true`.

File: src/app/exam/participants/Examinees.js [hunk 10]
```diff
@@ -10,75 +10,59 @@ import {TextareaModal} from "../../commons/modals/TextareaModal";
...
-            });
+    const addExaminees = (emails) => {
+        examService.addExaminees(examId, emails)
+            .then(errorList => console.log("errorList=", errorList)) // TODO: the error should be handled in the future
     }
 
     const removeExaminee = (email) => {
         examService.deleteExaminees(examId, email)
-            .then(() => removeExamineeByEmail(email));
+            .then(() => refetchExam(examId))
```
[the remark this answers is not in the document]
Comment (nyngwang) [reply src/app/exam/participants/Examinees.js:47]: 原本的`removeExamineeByEmail`只是前端資料在改, 直接重新fetch就可以拿到最新資料.

File: src/app/exam/questions/ExamQuestions.js [hunk 11]
```diff
@@ -12,95 +12,96 @@ import {useExamContext} from "./ExamContext";
...
-    const fetchExam = useCallback(() => {
-        examService.getExamOverview(examId).then(exam => {
-            exam.questions.sort((questionA, questionB) => questionA.questionOrder - questionB.questionOrder);
-            setQuestions(exam.questions);
-        });
-    }, [examId]);
+    const {url: currentURL} = useRouteMatch()
+    const {refetchExam, currentExam} = useExamContext()
+    const {examId} = useParams()
+    const NOT_SET = -1
```
[the remark this answers is not in the document]
Comment (nyngwang) [reply src/app/exam/questions/ExamQuestions.js:26]: rejudge功能會用到的常數

File: src/app/exam/questions/ExamQuestions.js (hunk 1 above)
Comment (nyngwang) [reply src/app/exam/questions/ExamQuestions.js:33]: 應該要是`undefined`, `null`是早期的寫法, 感謝指正.

File: src/app/exam/questions/ExamQuestions.js (hunk 2 above)
Comment (nyngwang) [reply src/app/exam/questions/ExamQuestions.js:29]: 因為`showRejudgeQuestionModal`存的值並不是`true/false`, 而是`int`, 用來表示哪個問題需要rejudge.

File: src/app/exam/score/ExamSummary.js (hunk 3 above)
Comment (nyngwang) [reply src/app/exam/score/ExamSummary.js:15]: 回 @Wally5077: `平均/總分`會比`平均`更有意義, 比如只顯示`100`不一定表示成績好, 因為總分可能是`1000`.

Comment (nyngwang) [reply src/app/exam/score/ExamSummary.js:15]: 回 @Chaoyuuu: 已在新的檔案接上API更新.

File: src/app/exam/score/ExamScore.js (hunk 5 above)
Comment (nyngwang) [reply src/app/exam/score/ExamScore.js:15]: 之前在line上問的UI修正, 已在新的檔案更新中移除, 感謝.

File: src/app/exam/participants/Examinees.js [hunk 12]
```diff
@@ -10,75 +10,59 @@ import {TextareaModal} from "../../commons/modals/TextareaModal";
...
     const {url: currentURL} = useRouteMatch();
     const {examId} = useParams();
-    const {currentExam, refetchExam} = useExamContext()
-    const [examinees, setExaminees] = useState(undefined);
-    const [selectedExaminee, setSelectedExaminee] = useState(undefined);
-    const [showAddStudentModal, setShowAddStudentModal] = useState(false);
-    const [showAddGroupModal, setShowAddGroupModal] = useState(false);
-
+    const {refetchExam, currentExam} = useExamContext()
+    const [examinees, setExaminees] = useState(undefined)
```
[the remark this answers is not in the document]
Comment (nyngwang) [reply src/app/exam/participants/Examinees.js:21]: 已在另一個留言回覆原因

File: src/app/exam/scoreboard/ExamScoreboardService.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    getExamScoreboard(examId) {
+        return Promise.all([
+            examService.getExaminees(examId),
+            examService.getExamOverview(examId),
+            examService.getExamScores(examId),
+        ])
+            .then(async ([examinees, exam, examScores]) => {
+                const out = {}
+                await Promise.all(exam.questions.map(async q =>
+                    problemService.getProblemById(q.problemId).then(problem => out[q.problemId] = problem.testcases.length)))
```
Comment (Wally5077) [inline src/app/exam/scoreboard/ExamScoreboardService.js:15]: problemService. getProblemsByIds(ids);

File: src/app/exam/submissions/ExamLiveSubmissions.js [hunk 14]
```diff
@@ -11,7 +11,7 @@ import VerdictIssuedEvent from "../../../models/VerdictIssuedEvent";
 import Verdict from "../../../models/Verdict";
 import './ExamSubmissions.scss';
 
-const ExamSubmissions = () => {
+const ExamLiveSubmissions = () => {
```
Comment (Wally5077) [inline src/app/exam/submissions/ExamLiveSubmissions.js:14]: 你這不要改名，我這邊會改

File: src/services/services.js [hunk 15]
```diff
@@ -2,11 +2,15 @@ import StudentService from "./StudentService";
...
+import {ExamScoreboardService} from "../app/exam/scoreboard/ExamScoreboardService";
+import {ExamScoreService} from "../app/exam/score/ExamScoreService";
 
 
 const studentService = new StudentService();
 const examService = new ExamService(studentService);
 const problemService = new ProblemService(studentService);
 const submissionService = new SubmissionService(studentService);
+const examScoreboardService = new ExamScoreboardService();
+const examScoreService = new ExamScoreService();
```
Comment (Wally5077) [inline src/services/services.js:14]: const examTranscriptService = new ExamTranscriptService(examService, problemService);

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: src/app/exam/participants/Examinees.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (nyngwang) [reply src/app/exam/participants/Examinees.js:47]: 想說用現有的`refetchExam(examId)`更新parent component, 這樣底下的相關component也會連帶更新, 就不寫多個函數做結果.一樣的事

File: src/app/exam/model.js [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+        this.examQuestions.forEach(question =>
+            this.#noSubByProblemId[question.problemId] = this.countNoSubByExamId(question.problemId))
+    }
+
+    countNoSubByExamId(problemId) {
+        return this.totalExaminees - this.#submissionsByProblemId(problemId).length;
+    }
+
+    get noSubByProblemId() {
+        return this.#noSubByProblemId
```
Comment (Wally5077) [inline src/app/exam/model.js:180]: 這檔案裡面有缺分號請一併補上

File: src/app/exam/scoreboard/ExamScoreboardPage.js [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+import {optionsTemplate, dataTemplate} from "./ChartConfigTemplates";
+import {examScoreboardService} from "../../../services/services";
+import {ExamScoreboardPresenter} from "./ExamScoreboardPresenter";
+
+
+const ExamScoreboardPage = () => {
+    const {url: currentURL} = useRouteMatch()
+    const {examId} = useParams()
+    const {refetchExam, currentExam} = useExamContext()
+    const [presenter, setPresenter] = useState(undefined)
```
Comment (Wally5077) [inline src/app/exam/scoreboard/ExamScoreboardPage.js:20]: 檔案內有缺分號的請一併補上

File: src/app/exam/scoreboard/ExamScoreboardPresenter.js [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+    get sumOfMaxScores() {
+        return this.#sumOfMaxScores
+    }
+
+    get totalExaminees() {
+        return this.#totalExaminees
+    }
+
+    //
```
Comment (Wally5077) [inline src/app/exam/scoreboard/ExamScoreboardPresenter.js:38]: 刪掉這行

File: src/app/exam/scoreboard/ExamScoreboardPresenter.js [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    constructor(examScoreboard) {
+        this.#examQuestions = examScoreboard.examQuestions
+        this.#crossAverage = examScoreboard.crossAverage
+        this.#sumOfMaxScores = examScoreboard.sumOfMaxScores
+        this.#totalExaminees = examScoreboard.totalExaminees
+        this.#testCasesDataByProblemId = examScoreboard.testCasesDataByProblemId
+        this.#totalPassDataByProblemId = examScoreboard.totalPassDataByProblemId
+        this.#totalTestCasesOf = examScoreboard.totalTestCasesOf
+        this.#questionAverageByProblemId = examScoreboard.questionAverageByProblemId
+        this.#countNoSubByProblemId = examScoreboard.noSubByProblemId
```
Comment (Wally5077) [inline src/app/exam/scoreboard/ExamScoreboardPresenter.js:23]: 檔案內檢查下分號問題

File: src/app/exam/scoreboard/ExamScoreboardService.js [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+            examService.getExaminees(examId),
+            examService.getExamOverview(examId),
+            examService.getExamScores(examId),
+        ]).then(async ([examinees, exam, examScores]) => {
+            const totalTestCasesOf = {}
+            // Currently `http://{{api}}/api/problems?ids=` won't return testcases info.
+            await Promise.all(exam.questions.map(question =>
+                problemService.getProblemById(question.problemId)
+                    .then(problem => totalTestCasesOf[problem.id] = problem.testcases.length)))
+            return {examinees, exam, examScores, totalTestCasesOf}
```
Comment (Wally5077) [inline src/app/exam/scoreboard/ExamScoreboardService.js:27]: 分號

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: src/services/ExamService.js [hunk 21]
```diff
@@ -27,6 +27,11 @@ export class ExamService extends AbstractService {
             .then(res => new Exam(res.data));
     }
 
+    async getExamScores(examId) {
+        return this.axios.get(`/api/exams/${examId}/transcript`)
+            .then(res => res.data);
+    }
```
Comment (Wally5077) [inline src/services/ExamService.js:33]: Method name change to 'getExamTranscript'

**Review by Wally5077 — CHANGES_REQUESTED** [review]

File: src/app/exam/scoreboard/ExamScoreboardService.js (hunk 13 above)
Comment (nyngwang) [reply src/app/exam/scoreboard/ExamScoreboardService.js]: 非常感謝你半夜趕工, 已修改!

File: src/app/exam/model.js [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+        this.#examinees = examinees;
+        this.#examQuestions = exam.questions.sort((q1, q2) => q1.questionOrder - q2.questionOrder);
+        this.#examScores = examScores;
+        this.#totalTestCasesOf = totalTestCasesOf;
+        this.#totalScoreByStudentId = {}
+        this.#studentScoresByStudentId = {}
+        this.#testCasesDataByProblemId = {}
+        this.#totalPassDataByProblemId = {}
+        this.#questionAverageByProblemId = {}
+        this.#noSubByProblemId = {}
```
[the remark this answers is not in the document]
Comment (nyngwang) [reply src/app/exam/model.js]: object要先初始化才能放值

File: src/app/exam/scoreboard/ExamTranscriptService.js [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+        return Promise.all([
+            this.examService.getExaminees(examId),
+            this.examService.getExamOverview(examId),
+            this.examService.getExamTranscript(examId),
+        ]).then(async ([examinees, examOverview, examTranscript]) => {
+            const totalTestCasesOf = {};
+            const ids = examOverview.questions.map(question => question.problemId);
+            await this.problemService.getProblemByIds(ids)
+                .then(problems => problems
+                    .forEach(problem => totalTestCasesOf[problem.id] = problem.testcases.length))
```
Comment (Wally5077) [inline src/app/exam/scoreboard/ExamTranscriptService.js:27]: 分號

File: src/app/exam/scoreboard/ExamTranscriptService.js [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+            return {examinees, examOverview, examTranscript, totalTestCasesOf}
```
Comment (Wally5077) [inline src/app/exam/scoreboard/ExamTranscriptService.js:28]: ditto

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

**Review by Johnny850807 — APPROVED** [review]
I think this can be merged first, as it's too long-live.

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

**Review by Wally5077 — APPROVED** [review]
LGTM

**Review by timt1028 — APPROVED** [review]
LGTM

**Review by Chaoyuuu — APPROVED** [review]
LGTM

**Pull request merged**